### ISSUE TITLE
Freeze during submodule open

### ISSUE THREAD

**Issue by mestihudson** [jesseduffield/lazygit#1298]
**Describe the bug**
When I select one of 6 submodules on Files/Submodules panel, lazygit freezes.

**To Reproduce**
1. open lazygit on terminal
2. navigate to Files/Submodules panel
3. press <enter> on selected submodule

**Expected behavior**
Navigate on specific submodule lazygit env 

**Screenshots**
`lazygit --logs shows` at time of crash.

<img width="1920" alt="Screen Shot 2021-04-26 at 14 19 22" src="https://user-images.githubusercontent.com/687967/116124318-8155f280-a69a-11eb-9ef0-f28d540bbb73.png">

**Desktop (please complete the following information):**
 - OS: MacOS Big Sur [v11.2.3]
 - Lazygit Version v0.28.1

**Comment by ablacklama** [comment]
I think this might be what i'm experiencing as well. I don't have any submodules in the repo but the command that causes the freeze is still showing the `--submodule` flag.
I can't `ctrl-C` or 'q' out of it. I just have to restart the terminal.

![image](https://user-images.githubusercontent.com/22246943/151722245-7fb33314-1407-4545-a285-8d29e2cb3805.png)


I'm running it through vscode's integrated terminal:
OS: Ubuntu 20.04
Lazygit Version v0.32.2

**Comment by jesseduffield** [comment]
Not sure what's happening there. We've got an integration test for selecting and entering a submodule. Would you be able to clone the lazygit repo and run:
```sh
go test pkg/gui/gui_test.go -run /submoduleEnter
```
to see if that passes on your end? That will help narrow down whether it's partially due to your environment

**Comment by jesseduffield** [comment]
(btw if it does pass, you can see how to make your own integration test in https://github.com/jesseduffield/lazygit/blob/master/docs/Integration_Tests.md which would be good for capturing a situation where it does fail)

**Comment by klement** [comment]
I'm seeing random slowdowns when entering a submodule (by pressing enter in UI). It's either immediate or takes a minute or two. Zero system load/CPU usage observed. I tried running the unit test but it's changed I gues...

```
ksekera@netgatedev ~/l/lazygit (master) [1]> go test pkg/gui/gui_test.go -run /submoduleEnter
cannot find package "." in:
        /home/ksekera/lazygit-src/lazygit/vendor/pkg/gui/gui_test.go
ksekera@netgatedev ~/l/lazygit (master) [1]> go test pkg/gui/gui.go -run /submoduleEnter
# command-line-arguments
pkg/gui/gui.go:73:24: undefined: fileWatcher
pkg/gui/gui.go:125:24: undefined: BackgroundRoutineMgr
pkg/gui/gui.go:209:14: undefined: ContextMgr
pkg/gui/gui.go:292:6: gui.resetHelpersAndControllers undefined (type *Gui has no field or method resetHelpersAndControllers)
pkg/gui/gui.go:294:16: gui.resetKeybindings undefined (type *Gui has no field or method resetKeybindings)
pkg/gui/gui.go:336:21: gui.contextTree undefined (type *Gui has no field or method contextTree)
pkg/gui/gui.go:358:22: undefined: NewContextMgr
pkg/gui/gui.go:464:6: gui.WatchFilesForChanges undefined (type *Gui has no field or method WatchFilesForChanges)
pkg/gui/gui.go:474:7: gui.createMenu undefined (type *Gui has no field or method createMenu)
pkg/gui/gui.go:481:16: undefined: guiCommon
pkg/gui/gui.go:481:16: too many errors
```

**Comment by stefanhaller** [comment]
> I'm seeing random slowdowns when entering a submodule (by pressing enter in UI).

I have seen the same occasionally. I also saw this sometimes when switching to a different repo using the recent repo menus. This especially happened when doing it quickly after starting lazygit. My suspicion was that there's maybe some mutex being held during a long-running operation (e.g. background fetch). I never investigated this more though.

**Comment by jesseduffield** [comment]
That's right: we wait for the fetch mutex to be free before we switch. It would be good if we could make our refresh commands cancelable so that switching repos is faster, or if we ensures that any writes they did have no effect by the time they're done, so that we can just let them finish without waiting for them
